[WIP] HDP changes

bhargavvader opened the pull request:
This is to address issues #262 , #901, #937, #945 - basically to attempt to clean up HDP as much as possible.

For starters, I've made HDP resemble `ldamodel` as much as possible - with regard to the imports, and the `dirichlet_expectation` method mirrors the `ldamodel` one.

Will keep making changes in this PR.

Edit: Would want to fix #952 with this as well.

bhargavvader commented:
@tmylk , I've also added a method for the HDP to LDA conversion, which returns an LDA model object.
Could you review that and the previous changes and see if it's fine?

tmylk commented on gensim/models/hdpmodel.py:
```diff
@@ -499,10 +502,21 @@ def hdp_to_lda(self):
+        Returns closest corresponding ldamodel object corresponding to current hdp model.
+        """
+        alpha, beta = self.hdp_to_lda()
+        ldam = ldamodel.LdaModel(num_topics=150, alpha=alpha, id2word=self.id2word)
```
why this number of topics?

bhargavvader replied:
It's 150 because the default number of topics in HDP before truncation is 150. This means all the internal matrices are of that size, and to make the `expElogbeta` matrice the same shape to accept the beta, we've to initialise `ldam` to that many topics.

tmylk replied:
Please add text to the doc-string about this choice

tmylk commented on gensim/models/hdpmodel.py:
```diff
@@ -34,11 +34,11 @@
 from __future__ import with_statement
 
 import logging, time
-import numpy as np
```
This is standard abbreviation. Please keep.

tmylk replied:
It will make reviewing this PR much easier if this decorative change wasn't here

bhargavvader replied:
Yeah, I'll make it back to `np`. 
But I still think it should be uniform throughout the package - will open another PR for the other 3.

bhargavvader commented:
@tmylk what sort of tests would you want to see? 
I removed the `np` to `numpy` change; PR #1000 works to make it uniform throughout. It should be easier to review this now though!

bhargavvader commented:
@tmylk could you have a look?

tmylk commented:
An example test: train HDP, create default LDA using the new method, check they are similar.

bhargavvader commented:
@tmylk , added tests, and a `random_state` method similar to LDA for more control during testing.

tmylk commented on gensim/models/hdpmodel.py:
```diff
@@ -49,20 +49,36 @@
+    return result.astype(alpha.dtype)  # keep the same precision as input
+
+
+def get_random_state(seed):
```
why duplicate https://github.com/RaRe-Technologies/gensim/blob/17a01cfa6363eb0c13600e4abcc7cfe68fd9c38d/gensim/models/ldamodel.py#L97 ?

bhargavvader replied:
Should I instead call it from the `ldamodel` class? How would this rather be done?

tmylk replied:
Better move to some shared class like util

tmylk replied:
sure

tmylk commented on gensim/test/test_hdpmodel.py:
```diff
@@ -51,12 +52,29 @@ class TestHdpModel(unittest.TestCase, basetests.TestBaseTopicModel):
-        self.model = self.class_(corpus, id2word=dictionary)
+        self.model = self.class_(corpus, id2word=dictionary, random_state=np.random.seed(0))
 
     def testShowTopic(self):
```
It's already in https://github.com/RaRe-Technologies/gensim/blob/9a3329cfccfcf98e9b88f5f52437f1049f4e3c44/gensim/test/basetests.py#L28

bhargavvader replied:
Yeah, but that only tests the types of the results; I thought it would be important to also test for particular values. The nature of my test is different. Should I maybe name it something else?

Also, how do I make this test file run the base class tests?

tmylk replied:
Yes, give these test a different name.
it should run base class test automatically. can you see them in the log?

bhargavvader replied:
Will do, and yup the `basetest` tests show up in HDP.

bhargavvader commented:
Changed the test name, moved functions to utils and matutils, and changed all the numpy to np there as well (I had forgotten to last time, only had changed the models).

bhargavvader commented:
@tmylk , could you please review? Travis fails on all the python versions because of 
`AttributeError: 'KeyedVectors' object has no attribute 'syn0'`, which is a `word2vec` (unrelated) test.

I've added `show_topic` as well now.

bhargavvader commented:
Ping @tmylk ! Any clue what to do with the tests?

tmylk commented:
Did keyed vectors get merged in correctly from develop branch?

bhargavvader commented:
How do I check if it has been? My guess is it hasn't, because the test is failing again. But when I do `git pull upstream develop` from the HDP branch I'm working on, it says everything is up to date.

bhargavvader commented:
@tmylk , I checked, and `syn0` is in the `keyedvector` class too. It's also identical to the file in the repo. Any clue why this is happening? And my changes shouldn't make the tests fail, anyway...

bhargavvader commented:
Ping @tmylk

bhargavvader commented:
@tmylk how do I check if keyedvectors is merged in correctly?

bhargavvader commented:
Sure, will close this and open 2 separate PRs this weekend.
